Create fresh _minimal_ CR for use with UpdateControl

tombentley opened the pull request:
### Type of change

- Bugfix

### Description

* Create fresh, _minimal_ CR for use with UpdateControl. This is what JOSDK requires in order to correctly make use of Server-Side Apply. See https://javaoperatorsdk.io/blog/2025/02/25/from-legacy-approach-to-server-side-apply/#resource-handling-without-and-with-ssa for more details.
* Add error reconciliation for KafkaServiceReconciler.
* Use consistent logging for when do a reconcile, or handle an error.


### Additional Context

Note: This clear up the errors being logged by the KafkaProxyReconciliation about getting an HTTP conflict when trying to patch the status.

_Why are you making this pull request?_

SamBarker commented on kroxylicious-operator/src/test/java/io/kroxylicious/kubernetes/operator/protocolfilter/KafkaServiceReconcilerTest.java:
```diff
@@ -70,7 +70,7 @@ void shouldMarkFilterAccepted() {
                                         assertThat(kafkaServiceStatus)
                                                 .singleCondition()
-                                                .hasObservedGenerationInSyncWithMetadataOf(input.getResource().get())
+                                                .hasObservedGenerationInSyncWithMetadataOf(kafkaService)
```
Oops this makes me notice we have KafkaServiceTest (at least but probably KafkaService) in the wrong package.

tombentley replied:
Yeah, I noticed that you put the KSReconciler in its own package. IDK whether that's the right thing to do or not. Probably it is, but I think we need to see how the Reconcilers interact with each other. 

Shall I move it into the same package as the other for now, just for consistency? And move them all later on when we know how they interrelate?

SamBarker replied:
I think we were expecting to put a few other things in the package so happy to move it now.

SamBarker commented on kroxylicious-operator/src/main/java/io/kroxylicious/kubernetes/operator/kafkaservice/KafkaServiceReconciler.java:
```diff
@@ -53,6 +77,18 @@ public List<EventSource<?, KafkaService>> prepareEventSources(EventSourceContext
     @Override
     public ErrorStatusUpdateControl<KafkaService> updateErrorStatus(KafkaService resource, Context<KafkaService> context, Exception e) {
-        return io.javaoperatorsdk.operator.api.reconciler.Reconciler.super.updateErrorStatus(resource, context, e);
+        var now = ZonedDateTime.ofInstant(clock.instant(), ZoneId.of("Z"));
```
This is clunky, I'm almost tempted to say we should have our own clock wrapper to do this for us so can just say `clock.zonedInstant()` or something. Nothing to change in this PR however.

tombentley replied:
Agree on both points.

SamBarker replied:
https://github.com/kroxylicious/kroxylicious/pull/1985

SamBarker commented on kroxylicious-operator/src/main/java/io/kroxylicious/kubernetes/operator/kafkaservice/KafkaServiceReconciler.java:
```diff
@@ -36,14 +48,26 @@ public UpdateControl<KafkaService> reconcile(KafkaService resource, Context<Kafk
+
+    private static KafkaService newServiceWithCondition(KafkaService resource, Condition acceptedCondition) {
+        // @formatter:off
+        final KafkaService amended = new KafkaServiceBuilder()
```
are we assigning to `amended` just so we can turn the formatter on and off? Or can it be inlined now its not being passed directly to `patchStatus`

tombentley replied:
I can inline it.

Another thing I noticed is that this reconciler names the `KafkaService` parameter `resource`. The `KafkaProxyReconciler` calls it `primary`. The rest name the parameter after the type (so like `filter` or `ingress`). We should pick one and stick to it. I think I prefer naming it after the type because I don't really find the `primary` terminology particularly helpful, and all the things (primaries, secondaries and dependent) are resources.

SamBarker replied:
Naming after the type probably makes the most sense.

SamBarker commented on kroxylicious-operator/src/main/java/io/kroxylicious/kubernetes/operator/KafkaProxyReconciler.java:
```diff
@@ -140,8 +140,13 @@ private static KafkaProxy buildStatus(KafkaProxy primary,
         // @formatter:off
-        return new KafkaProxyBuilder(primary)
-                .editOrNewStatus()
+        return new KafkaProxyBuilder()
+                .withNewMetadata()
+                    .withName(ResourcesUtil.name(primary))
+                    .withNamespace(ResourcesUtil.namespace(primary))
+                    .withUid(ResourcesUtil.uid(primary))
+                .endMetadata()
+                .withNewStatus()
```
Could this be simplified along the lines of 
```suggestion
        return new KafkaProxyBuilder()
               .withNewMetadataLike(primary.getMetadata())
                .endMetadata()
                .editOrNewStatusLike(primary.getStatus())
```

Modulo formatting...

tombentley replied:
No, because `withNewMetadataLike(primary.getMetadata())` will include all the other metadata in the primary, which we explicitly don't want to copy. We don't want to be "field manager" for all the labels, for example, because that would prevent users using `kubectl` to set those labels. Other fields are really owned by the apiserver, like the `resourceVersion` (TBH I don't know exactly _how_ the apiserver behaves if you include those in a SSA patch, but that doesn't change that we shouldn't be trying to set them in a patch).

SamBarker replied:
Ok, thats cool. 

I might have a poke at adding a method on ResourceUtil to produce a co-ordinates metadata based on a resource.

SamBarker replied:
https://github.com/tombentley/kproxy/pull/20

tombentley replied:
> I might have a poke at adding a method on ResourceUtil to produce a co-ordinates metadata based on a resource.

tombentley replied:
As in, I'm about to push something along the lines I mentioned this morning.

SamBarker commented on kroxylicious-operator/src/main/java/io/kroxylicious/kubernetes/operator/KafkaProxyIngressReconciler.java:
```diff
@@ -84,13 +87,20 @@ public UpdateControl<KafkaProxyIngress> reconcile(
     private static KafkaProxyIngress newIngressWithCondition(KafkaProxyIngress ingress, Condition condition) {
         // @formatter:off
-        return new KafkaProxyIngressBuilder(ingress)
+        return new KafkaProxyIngressBuilder()
```
similar thought about `withNewMetadataLike`

tombentley replied:
Same answer.

SamBarker reviewed (approved):
LGTM

SamBarker commented on kroxylicious-operator/src/main/java/io/kroxylicious/kubernetes/operator/KafkaProtocolFilterReconciler.java:
```diff
@@ -176,7 +176,9 @@ public UpdateControl<KafkaProtocolFilter> reconcile(
         LOGGER.debug("Patching with status {}", newFilter.getStatus());
         UpdateControl<KafkaProtocolFilter> uc = UpdateControl.patchStatus(newFilter);
-        LOGGER.info("Completed reconciliation of {}/{}", namespace(filter), name(filter));
+        if (LOGGER.isInfoEnabled()) {
```
Is sonar really whinging about stuff logged at info?!

tombentley replied:
It was indeed.

tombentley commented on kroxylicious-operator/src/main/java/io/kroxylicious/kubernetes/api/common/Condition.java:
```diff
@@ -108,6 +107,18 @@ public void setReason(String reason) {
+    public String toString() {
+        return "Condition{" +
+                "observedGeneration=" + observedGeneration +
+                ", type=" + type +
+                ", status=" + status +
```
Having the type and status first is pretty helpful when assertions fail.

tombentley commented on kroxylicious-operator/src/main/java/io/kroxylicious/kubernetes/operator/KafkaServiceReconciler.java:
```diff
@@ -0,0 +1,63 @@
+ * Licensed under the Apache Software License version 2.0, available at http://www.apache.org/licenses/LICENSE-2.0
+ */
+
+package io.kroxylicious.kubernetes.operator;
```
Not sure why the file rename has come through as a create and file.

tombentley commented:
@SamBarker I ended up getting lured down the path of repeated factoring out copy-pasted code. So now there are some more utility functions in the `ResourcesUtil`:

* For creating `Conditions` of different types and statuses.
* Creating a patch resource (for three of the CRs) with an updated status.condition. This preserves existing conditions of other types,. It also won't overwrite a condition which has a (strictly) higher generation (which shouldn't happen, but there seems no harm in doing the right thing in this situation).
    * I didn't do this for KafkaProxyReconciler because I think think this is going to get heavily refactored soon, as we introduce the VKCReconciler. 
* I also moved KafkaServiceReconciler and its test to the operator package, and renamed the parameter, as discussed.

SamBarker commented on kroxylicious-operator/src/test/java/io/kroxylicious/kubernetes/operator/ResourcesUtilTest.java:
```diff
@@ -173,4 +179,78 @@ void toLocalRef() {
+        );
+    }
+
+    @ParameterizedTest(name = "{0}")
+    @MethodSource
+    void maybeAddOrUpdateCondition(String testName, List<Condition> list, Condition condition, List<Condition> expectedResult) {
+        assertThat(ResourcesUtil.maybeAddOrUpdateCondition(list, condition)).isEqualTo(expectedResult);
+    }
```
using [`ArgumentSet`](https://junit.org/junit5/docs/5.11.0/api/org.junit.jupiter.params/org/junit/jupiter/params/provider/Arguments.ArgumentSet.html) instead of just `Arguments` means you can avoid the first arg here and the need to add `name = "{0}"`

tombentley replied:
Fixed. Thanks!

SamBarker commented on kroxylicious-operator/src/main/java/io/kroxylicious/kubernetes/operator/ResourcesUtil.java:
```diff
@@ -224,4 +241,168 @@ static <O extends HasMetadata, R extends HasMetadata> Set<ResourceID> findReferr
+        return newTrueCondition(clock, observedGenerationSource, Condition.Type.ResolvedRefs);
+    }
+
+    static Condition newResolvedRefsFalse(Clock clock,
```
I'm not sold on including the true and false in the method names. Looking just at the usages in this diff I think we could just as easily use `newResolvedRefs` passing the true or false flag from the if conditions which decide which method to call.

tombentley replied:
The difference between true and false is not just the `status`, it's also the presence of the `reason` and `message` (`false` has them, `true` does not). Likewise for `unknown`, which we set when `reconcile()` throws, we're taking the reason and message from the exception. 
So, sure, we _could_ have a single method that took the `status` as an argument, and nullable `reason` and `message`, but then:
* we either don't validate the rules about the presence of those things, or we'd have to write the code to check that at runtime. 
* we'd have to push the how we get the reason and status for unknown into the call sites. 
Overall, I felt those were worse than having dedicated factory methods for the three status cases.

I think there's a stronger case for saying the `Type` should be given as an argument.

SamBarker replied:
Hmm I take the point about reason and message. Part of the problem is `newResolvedRefsFalse ` doesn't really mean anything to me. Maybe something like `newUnresolvedRefs[Condition]` or `newMissingRefs[Condition]` or something else that explains why its false (I'm ok with that delegating the actual construction to `newFalseCondition`)

tombentley replied:
Personally I prefer the names that map more directly to the cases of the `Status` enum, and which also clearly associate with the `type`. The trouble with having method like `newUnresolvedRefs` and `newMissingRefs` is that those could be different types of condition entirely, but that's not the `Condition` model we have.

SamBarker replied:
it could be different condition types entirely yes, and changing that shouldn't ripple half way through the code base. We are tying the logical role to the implementation details.

SamBarker commented on kroxylicious-operator/src/main/java/io/kroxylicious/kubernetes/operator/ResourcesUtil.java:
```diff
@@ -224,4 +241,168 @@ static <O extends HasMetadata, R extends HasMetadata> Set<ResourceID> findReferr
+    }
+
+    static ConditionBuilder newConditionBuilder(Clock clock, HasMetadata observedGenerationSource) {
+        var now = ZonedDateTime.ofInstant(clock.instant(), ZoneId.of("Z"));
```
This does mean that every condition is created with its own time down to some milli/nano second level. Is that what we want or do we want a single time that represents the reconciler pass?

tombentley replied:
It does, but _right now_ each status only every has a single condition, so there's only a single `lastTransitionTime`. I think we can re-evaluate when we have a Reconciler which is setting two conditions in the same call to `reconcile()`.

SamBarker replied:
I find my self swinging both ways on this. 

1. If the intent is to have a single time across the reconciliation cycle then we should pass it in as an integral part of the API.
2. It will be generated exactly the same way at every calls site so we should inline it... 

I worry that if its kept down in this method we are likely to forget to re-visit it when we actually do have a reconciler with multiple conditions (but maybe we never do??).

tombentley replied:
We're passing the `clock` around, so it's not like there's a call to `System.currentTimeMillis` lurking in there trying to be forgotten. I think the `clock` will be enough to remind us in that glorious future time of Reconcilers managing multiple Conditions.

SamBarker commented on kroxylicious-operator/src/main/java/io/kroxylicious/kubernetes/operator/ResourcesUtil.java:
```diff
@@ -224,4 +241,168 @@ static <O extends HasMetadata, R extends HasMetadata> Set<ResourceID> findReferr
+        if (conditionsOfType.isEmpty()) {
+            conditionsOfType.add(condition);
+        }
+        else if (conditionsOfType.get(0).getObservedGeneration() <= condition.getObservedGeneration()) {
+            conditionsOfType.set(0, condition);
+        }
```
This means that if our condition has a generation that is older than the current highest it gets left on the floor. Should we sort it into position in the list? Probably not as it sort of implies we are acting on stale data so maybe we should the throwing an exception in that case to abort the reconciliation pass?

tombentley replied:
This shouldn't happen. It would be unambiguously incorrect to replace a condition with a higher generation with the given one. So either we drop the given condition, we grow the list, or we throw an exception.

* I don't think it helps to keep multiple conditions of the same type around. It gives no useful or actionable information to the user (let's not forget that `status` is our API to a user). We'd still need some logic to prune them, and that's gonna be difficult to test precisely because this should never happen.
* I don't think throwing is ideal either. That will manifest as a weird exception in the `status`, which the user likely won't understand or know how to fix. They might not even be _able_ to fix the problem. And whatever the cause is, it's very likely not the operator that's at fault.

SamBarker replied:
>I don't think throwing is ideal either. That will manifest as a weird exception in the status, which the user likely won't understand or know how to fix. They might not even be able to fix the problem. And whatever the cause is, it's very likely not the operator that's at fault.

I don't think it should manifest in the status but I do think its a single that something has gone wrong and the reconciler should abort what ever it currently has in progress as its acting on stale data (isn't that the implication?). 

I think it means in the best case the operator is acting on stale data, at worst doesn't it imply that the operator is screwing something up by messing with generations? OHh could it be sign that there are multiple copies of the operator running (I don't think thats something we intend to support)

tombentley replied:
> I think it means in the best case the operator is acting on stale data, at worst doesn't it imply that the operator is screwing something up by messing with generations? OHh could it be sign that there are multiple copies of the operator running (I don't think thats something we intend to support)

It cannot occur due to stale data because the apiserver updates to a resource should be atomic and cover the whole resource. Whatever it later serves up (via a GET or a WATCH) will have a consistent `metadata`, `spec` and `status`. And the operator is updating the `status` based on the `spec` in the same request.

_If_ it happened it would imply that _something_ was messing with the `status.generation` (i.e. creating a `status` with a `status.observedGeneration` that was greater than the `metadata.generation`). The problem is we can't assume that that something is _this_ (or _any_) instance of the operator. 

Assume for a moment that you're right and there is some errant process somewhere. Suppose it eventually gets killed. What is this instance of the operator going to do _then_? What you're arguing for is for it to continue throwing an exception, because it will still be true that the `observedGeneration` on the `status` is greater than expected. I.e. the operator won't self-heal. Someone will have to fix it, if they can figure out what they need to do from the exception. It seems that `kubectl` now supports editing the `status`, so I suppose fixing it is easier that it once was, but it's that's a relatively recent addition. I really don't think that throwing in this situation is making anyone's life better.

SamBarker replied:
but this change detects that we did a bunch of work and that something is very screwy  and silently ignores that fact. Is it safe to assume that any changes we are trying to apply are actually safe or sensible to apply if something else is messing with the CR?
